hal/ia32/timer.c

```diff
@@ -49,56 +50,95 @@
 #define HPET_CONFIG_TMR_PERIODIC_CAN_SET (1u << 6)
 #define HPET_CONFIG_TMR_32BIT_MODE       (1u << 8)
 
-typedef enum { timer_unknown, timer_pit, timer_lapic, timer_hpet } timerType_t;
+typedef struct {
+	enum { timer_undefined, timer_pit, timer_lapic, timer_hpet } type;
+	unsigned int (*name)(char *s, unsigned int *len);
+	int (*init)(u32 intervalUs);
+
+	/* When used as scheduling timer */
+	int (*schedulerIrq)(unsigned int, cpu_context_t *, void *);
+	void (*schedulerSetWakeup)(u32 waitUs);
+	void (*schedulerInitCore)(unsigned int id);
+
+	/* When used as timestamp timer */
+	time_t (*timestampGetUs)(void);
+	time_t (*timestampBusyWaitUs)(time_t waitUs);
+} hal_timer_t;
+
 
 struct {
 	intr_handler_t handler;
 	spinlock_t sp;
 	u32 intervalUs;
 
-	timerType_t schedulerTimerType;
-	timerType_t timestampTimerType;
-	union {
-		struct {
-			volatile time_t jiffies;
-		} pit;
-		struct {
-			volatile u64 cycles; /* How many ticks were there on CPU0 */
-		} lapic;
-		struct {
-			hal_gasMapped_t addr;
-			u32 period;
-			intr_handler_t tmr1;
-		} hpet;
-	} timestampTimer;
-	union {
-		struct {
-			u32 frequency;
-			u32 wait[MAX_CPU_COUNT]; /* Wait time (in cycles) of this CPU */
-		} lapic;
-	} schedulerTimer;
-} timer;
+	const hal_timer_t *schedulerTimer;
+	const hal_timer_t *timestampTimer;
 
+	struct {
+		volatile time_t jiffies;
+	} pitData;
 
-int hal_timerRegister(int (*f)(unsigned int, cpu_context_t *, void *), void *data, intr_handler_t *h)
+	struct {
+		volatile u64 cycles; /* How many ticks were there on CPU0 */
+		u32 frequency;
+		u32 wait[MAX_CPU_COUNT]; /* Wait time (in cycles) of this CPU */
+	} lapicData;
+
+	struct {
+		hal_gasMapped_t addr;
+		u32 period;
+		intr_handler_t tmr1;
+	} hpetData;
+} timer_common;
+
+
+static unsigned int _hal_sprintU32(char *s, unsigned char len, u32 n)
```

Comment:
One more function to convert `integer` to `string`.
Why not `hal_i2s()`?

Reply:
Right, it's even exported to hal API now too